Holo docs for alpha (#405)
holochain/docs-pages: src/pages/references/enable-holo.md

---
title: Enable Apps for Holo Hosting
---

## Intro

Holo is cloud hosting for Holochain. Holo helps you enable your users to experience your Holochain application with the simplicity of a web browser and without maintaining a node.

A core distinction versus running nodes on the traditional cloud is that Holo provides the key management and signing infrastructure for users' keys to be held client-side, so your users always have full agency.

## Get started

Holo hosting is for Holochain applications. It is assumed that you have familiarity with Holochain.

https://developer.holochain.org/get-started/

There are two pathways in this section --- scaffolding a new hApp or migrating an existing hApp. Both pathways contain largely the same information, so choose the pathway that best applies to your situation.

> In both cases there are no required DNA changes for Holo Hosting --- only UI changes. However, Holo operates under a different context and we highly recommend that you read the core concepts, in particular the sections on membrane proofs and anonymous access, as these **will need DNA integrity zome changes to enable certain features**.

Author: Ah! I meant the Holo Core Concepts which is the section below that you're referring to. I wonder how we structure this better

```shellsession
cd super_todos
```

Now enter the nix development shell (which makes all scaffolding tools as well as the Holochain binaries directly available from the command line) with:

```shellsession
nix develop
```

You should see:

::: output-block
```text
Holochain development shell spawned. Type exit to leave.
```
:::

Finally, install the `npm` dependencies with:

```shellsession
npm install
```

### UI

The biggest difference between a Holo and pure Holochain application is in the UI. The scaffolding tool should automatically have generated everything required here, so we'll simply point out the primary differences.

In a Holo setting we use the [`@holo-host/web-sdk`](https://www.npmjs.com/package/@holo-host/web-sdk) library instead of `@holochain/client`. WebSDK provides an instance of `AppAgentWebsocket`, which is almost the same as `AppWebsocket`.

> ```typescript

Author: It should be in App.vue. How should we add it in? Do we just spell it out like "Let's look at App.vue"? Or is there a smarter way?

```
:::

The final step is create a collection that can be used to render all of to-do items that users create. To create a collection, type:

```shellsession
hc scaffold collection
```

You should then see:

::: output-block
```text
Collection name (snake_case, eg. "all_posts"): ›
```
:::

Enter `my_todos` and press <kbd>Enter</kbd>. You should then see:

::: output-block
```text
? Which type of collection should be scaffolded? ›
❯ Global (get all entries of the selected entry types)
  By author (get entries of the selected entry types that a given author has created)
```
:::

Use the arrow key to select `By author` and press <kbd>Enter</kbd>. You should then see:

::: output-block
```text
? Which entry type should be collected? ›
❯ TodoItem
```
:::

Press <kbd>Enter</kbd>. You should then see:

::: output-block
```text
Collection "my_todos" scaffolded!
```
:::

Holo strongly recommends that you implement a [**membrane proof**](/glossary/#membrane-proof) for production or long-running applications. However, membrane proofs that restrict read-only access prevent Holo from providing always-on nodes and [read-only access to applications](https://link.tbd).

> If you want to benefit from always-on nodes and read-only access while having membrane proofs, you will need to implement special "Holo-safe" logic. You can [read more here](https://link.to.core.concepts).

Reviewer: I'd like to have a Holo core concepts section, sure. Then we could link out to it from all sorts of places.

### Testing

Holo provides the `holo-dev-server` binary, which simulates the Holo network locally for development. `holo-dev-server` serves a copy of Chaperone, the JavaScript library that connects the browser to the Holo network, and runs a Holochain conductor. Like the real Holo network, `holo-dev-server` uses `.happ` bundles, which do not include a UI. (The `hc` developer tool creates both a `.happ` bundle and a `.webhapp` bundle for you.) When you scaffold an app with the `--holo` flag, its dev environment will provide `holo-dev-server` on the command line for you.

1. Ensure you are in the nix development environment by checking the shell prompt. Run `nix develop` in the root of your project's folder if you aren't.
2. Run `npm start:holo`, which:
    1. Starts `holo-dev-server' and provisions instances (cells) of your DNA for two agents,
    2. Automatically opens an instance of a network inspector called [Holochain Playground](https://github.com/darksoil-studio/holochain-playground) in your browser, and

>     3. Runs a web server to serve the UI.

Author: Yea I guess technically. It's an out-of-the-box thing for most UI frameworks to make the UI available at localhost right?

>     3. Runs a web server to serve the UI.

Reviewer: ohhhhhhhhhhh it's the UI framework itself that has a local dev server? got it.

>     3. Runs a web server to serve the UI.

Author: Ah I see. Yea

> 3. Access your two agents on your browser at `http://localhost:8888` and `http://localhost:8889`.

Author: At launch, they are two separate anonymous instances with throwaway keys. Depending on application logic they might just immediately direct to a login/signup page, or allow read-only use at first.

I don't know exactly how we keep their chaperone instances separate.

The windows should not be very exciting yet, because you haven't edited the hApp to use the generated UI elements, but what you see on the screen should be some hints on how to proceed.

So let's follow those hints. Switch to a code editor for these steps:

1. Open the Vue file `ui/src/App.vue` and add two imports near the top, just below `<script lang="ts">`:

    ```typescript
    import MyTodos from './todos/todos/MyTodos.vue';
    import CreateTodoItem from './todos/todos/CreateTodoItem.vue';
    ```

2. Replace the "EDIT ME" text there with:

    ```html
    <CreateTodoItem></CreateTodoItem>
    <MyTodos author={client.myPubKey}></MyTodos>
    ```

Save the file, and you should see that the windows have auto-updated to show something that should look like this:

![Two web browser windows showing basic widgets to manipulate the shared to-do list](assets/img/scaffolded_todos.png)

You now have a fully functional Holo app up and running!

## Migrate from a pure Holochain app

> Only minor UI changes are technically required for Holo hosting. However, Holo operates under a different set of assumptions and we highly recommend that you read the [Core Concepts](/concepts/). It is likely that you will want to reconsider certain UX flows.

Author: Yup - let's agree how we want to structure the Holo "core concepts" and what name we even want to call it, given the confusion with Holochain Core Concepts

This example will use an example forum hApp with a Vue-based UI, but feel free to follow along with your own hApp instead.

```shellsession
nix run github:holochain/holochain#hc-scaffold -- example forum
```

In a Holo setting we use [`@holo-host/web-sdk`](https://www.npmjs.com/package/@holo-host/web-sdk) which provides an instance of `WebSDK`, instead of `AppAgentWebsocket` from `@holochain/client`. Both `WebSDK` and `AppAgentWebsocket` implement `AppAgentClient`.

You'll first need to ensure that you use `WebSDK` or `AppAgentWebsocket` depending on the context:

```typescript
import { AppAgentClient, AppAgentWebsocket } from '@holochain/client';
import WebSdk from '@holo-host/web-sdk'; // Import WebSDK library

export default defineComponent({
  components: {
    //...
  },
  data(): {
    client: AppAgentClient | undefined; // Ensure that client is not AppAgentWebsocket
    //...
  } {
```

> You can then initialize a context-dependent client, assuming you want your UI to be compatible with both Holo and Holochain.

Author: The data() function itself doesn't check right?

```typescript
  data(): {
    client: AppAgentClient | undefined;
    loading: boolean;
    IS_HOLO: boolean;
  } {
    return {
      client: undefined,
      loading: true,
      IS_HOLO: ['true', '1', 't'].includes(import.meta.env.VITE_APP_IS_HOLO?.toLowerCase()), // You should set this as an env variable
    };
  },

>   async mounted() {

Author: mounted()? This is a Vue lifecycle function to specify logic that occurs when the UI elements are mounted to the DOM. And yes we add logic in here specific to Holo.

I haven't had all the experience of writing documentation that you have so I'm not sure what we really need to call out about this - seemed self-apparent to me. Can you suggest something?

>   async mounted() {

Reviewer: if the function already exists in the Vue file (say, as a stub) let's just be explicit -- "edit the two Vue lifecycle functions so they read as follows" or something like that.

    if (this.IS_HOLO) {
      const client: WebSdk = await WebSdk.connect({
        chaperoneUrl: import.meta.env.VITE_APP_CHAPERONE_URL, // We'll explain this in the testing section
        authFormCustomization: {
          appName: 'forum-app', // Display name on the credentials form. You can also set it in Cloud Console when deploying
        }
      });

      this.client = client
    } else {
      this.client = await AppAgentWebsocket.connect('', 'vue-default');
      this.loading = false;

>     }

Author: ```suggestion
    }
}
```

> ```

Author: Yes. I didn't think it was important but you're right that it's literally the only thing left out

The WebSDK client loads and connects to an iframe called Chaperone. Chaperone is a key and connection manager to the Holo network and is invisible outside a few specific cases.

Because the Holo network acts like a remote conductor, you'll need to handle connectivity issues. Changes to the connection are emitted and you can register an event handler if the client is an instance of WebSDK. You can see the full list of emitted events here: https://github.com/Holo-Host/web-sdk

> ```typescript

Author: Yea...structurally it's part of the snippet above actually. I manually split it out but perhaps that's a bad idea that causes confusion. I'll make a separate commit to address this.

import type { AgentState } from '@holo-host/web-sdk';

async mounted() {
  if (this.IS_HOLO) {
  // ...

    // register event handler for agent-state.
    client.on('agent-state', (agent_state: AgentState) => {
      this.loading = !agent_state.isAvailable
    });

    // ...

> ```

Author: ```suggestion
    }
}
```
```

Reviewer: annoyingly, the way GH's suggestion fences work, the closing fence is lost. I ran into this yesterday -- had to edit it locally and push the changes.

Reviewer: I looked at the actual current state and it looks like you resolved this 👍🏼

Now your UI can connect to the Holo network and read forum posts. However, users can't post or comment yet because they don't yet have a source chain. They are considered "anonymous" and can only request public data from the network.

!!! note

> Unlike in pure Holochain, you cannot assume that a user has provided their desired keys when the connection is established. This may require changes to your application logic. You can read more at https://link.to/keys/documentation

Author: Yes they may not have generated their keys yet. When they haven't, we have some throwaway / temporary keys generated for them. To be honest I'm unclear as to why.

!!!

Holo uses "sign-up" and "sign-in" terminology but there is no external authentication process. In both processes a user (re)derives their keys from email and password, with sign-up also triggering the installation of new hApp cells on a hosting device in the Holo network.

So let's add sign-up and sign-in functionality to the app.

> ```html

Author: It's still part of App.vue in the scaffold I think? Just in the html section.

<main>
  <h1>Forum</h1>
  <!--Add these buttons-->
  <div v-if="IS_HOLO && !isLoggedIn">
    <mwc-button
      style="margin-top: 16px"
      raised
      label="Sign Up"
      @click="signUp"
    />

    <mwc-button
      style="margin-top: 16px"
      raised
      label="Sign In"
      @click="signIn"
    />
  </div>

  <div id="content">
    <h2>All Posts</h2>
    <AllPosts></AllPosts>
    <span style="margin-bottom: 16px"></span>
    <CreatePost></CreatePost>
  </div>
</main>
```

> ```typescript

Author: App.vue again. Lower down after the data() section there's a methods section.

> ```typescript

Reviewer: okay, so it sounds like all the edits are to `App.vue` then? Let's put something near the top to say "all edits you'll be making will be to your UI's entry point -- in this case, `App.vue`" and then give a bit of context for each snippet (e.g., "put this stuff in the section that starts with ___")

methods: {
  // ...
  async signUp () {
    await (this.client as WebSdk).signUp({ cancellable: true });
  },

  async signIn () {
    await (this.client as WebSdk).signIn({ cancellable: true });
  },

  // ...
}
```

These trigger the display of a full screen credentials modal from the Chaperone iframe, defaulting to either a sign-up or sign-in form. The user can also toggle between them directly.

![](https://hackmd.io/_uploads/r1WJbrkC3.png)

> During sign-up a registration code field can be shown to enable [**membrane proof**](/glossary/#membrane-proof) workflows. Holo strongly recommends that you implement a membrane proof for production or long-running applications. However, membrane proofs that restrict read-only access prevent Holo from providing always-on nodes and [read-only access to applications](https://link.tbd).

Author: Thanks. Yes it's good to keep the info aligned across both.

Finally, we will also need sign-out functionality to clear user keys from local storage.

```html
<main>
  <h1>Forum</h1>
  <!--Add these buttons-->
  <div v-if="IS_HOLO && !isLoggedIn">
    <!-- ... -->
  </div>
  <div v-else>
    <mwc-button
      style="margin-top: 16px"
      raised
      label="Sign Out"
      @click="signOut"
    />
  </div>
  <!-- ... -->
</main>
```

```typescript
methods: {
  // ...
  async signOut () {
    await (this.client as WebSdk).signOut());
  },

  // ...
}
```

Now you're ready to build and test!

### Building and testing

Holo provides the `holo-dev-server` binary, which simulates the Holo network locally for development. `holo-dev-server` serves a copy of Chaperone, the JavaScript library that connects the browser to the Holo network, and runs a Holochain conductor. Like the real Holo network, `holo-dev-server` uses `.happ` bundles, which do not include a UI. (The `hc` developer tool creates both a `.happ` bundle and a `.webhapp` bundle for you.)

To include `holo-dev-server` in your development environment, update your project's `flake.nix` file to the below, and re-enter the shell:

> ```nix

Author: We can't as Holo Network itself is on 0.2.x

I'm happy to keep this as a feature branch until we get to 0.2.x stable

> ```nix

Reviewer: aw dang! Good to know. Yes, let's keep this as a feature branch unless we get a pull (from you, Rob, some business decision-maker, whoever) to get that documentation online. We'll want to put a caveat at the top that hApp devs will need to either start scaffolding from some 0.2 channel of Holonix or upgrade their DNAs to 0.2.

{
  description = "Template for Holochain app development";

  inputs = {
    nixpkgs.follows = "holochain/nixpkgs";
    holochain = {
      url = "github:holochain/holochain";
      inputs.versions.url = "github:holochain/holochain/?dir=versions/0_2";
    };
  };

  outputs = inputs @ { ... }: let
    holoDevServerChannel = "1774"; // TODO: Change to alpha once merged in
  in
    inputs.holochain.inputs.flake-parts.lib.mkFlake {
      inherit inputs;
    } {
      systems = builtins.attrNames inputs.holochain.devShells;
      perSystem = { config, pkgs, system, ... }: {
        devShells.default = pkgs.mkShell {
          inputsFrom = [ inputs.holochain.devShells.${system}.holonix ];
          packages = with pkgs; [ nodejs-18_x curl ];

          extraSubstitutors = [ "https://cache.holo.host" ];
          trustedPublicKeys = [
            "cache.holo.host-2:ZJCkX3AUYZ8soxTLfTb60g+F3MkWD7hkH9y8CgqwhDQ="
          ];

          shellHook = ''
            nix-env -f "https://hydra.holo.host/channel/custom/holo-nixpkgs/${holoDevServerChannel}/holo-nixpkgs/nixexprs.tar.xz" -iA holo-dev-server
          '';
        };
      };
    };
}
```

> To connect your application to `holo-dev-server`, point the client's `chaperoneUrl` to `https://localhost:24274`.

Author: Ok any action required here? If not please hit resolve :)

> To connect your application to `holo-dev-server`, point the client's `chaperoneUrl` to `https://localhost:24274`.

Reviewer: I have no idea why I've got it as a suggestion that changes nothing 🤔 I'll think of a way to elaborate on what's needed.

```json
  "scripts": {
    // ...
    "build:happ": "npm run build:zomes && hc app pack workdir --recursive",
    "build:zomes": "RUSTFLAGS='' CARGO_TARGET_DIR=target cargo build --release --target wasm32-unknown-unknown"
  },
```

Let's add the following scripts to the top of the `"scripts"` object. These will:

1. Build your `.happ` file,
2. Run `holo-dev-server` with your `.happ` file,
3. Run the `holochain-playground` at `localhost:4444`, and
4. Start the your UI at `localhost:8888`.

```json
    "start:holo": "AGENTS=2 npm run network:holo",
    "network:holo": "npm run build:happ && UI_PORT=8888 concurrently \"npm run launch:holo-dev-server\" \"holochain-playground ws://localhost:4444\" \"concurrently-repeat 'VITE_APP_CHAPERONE_URL=http://localhost:24274 VITE_APP_IS_HOLO=true npm start -w ui' $AGENTS\"",
    "launch:holo-dev-server": "sudo ./holo-dev-server workdir/test-app.happ",
```

These scripts assume that you have a `start` script setup in a `ui` workspace, which you should have if you created the hApp using our scaffolding tool.

You now have a fully functional Holo app up and running!

> !!! info

Author: Do you mean the chaperoneUrl advice? I'm sure there is. We can add it to the todo list

> !!! info

Reviewer: yes, the chaperoneUrl advice, along with `VITE_APP_IS_HOLO` env var. So you've got three scenarios:

* `build` w/o any env vars builds the UI for native HC
* `build:holo` passes `VITE_APP_IS_HOLO=true` and a production `VITE_APP_CHAPERONE_URL` env vars (with alternatives for other front-end toolkits, of course)
* for local dev, allow the parent `package.json` to specify those env vars, as in the example you gave.

For production deployment, make sure that `chaperoneUrl` is configured to `https://chaperone.holo.hosting`
!!!

## Deploy

This process assumes you already have a Cloud Console account. If you do not, you can sign up at https://register.holo.host

### Deploy UI

Holo does not currently offer UI hosting. hApp managers will need to deploy their UI separately for now.

!!! info
Holo supports "headless hosting" where there is no Holo-enabled UI. In this case Holo will maintain copies of the DHT, but will not hold any source chains. Holo hosting requires <membrane proof recommendation?>

> Holo does not support direct programmatic/API access to deployed hApps.

Author: It's just not productised yet. We actually have a POC implementation of this in our HRL gateway that fractal-tribute uses.

!!!

### Deploy hApp to Holo network

1. Log into Cloud Console.
2. Click on "Add a hApp".
3. Fill out hApp details:
    * "Link to hApp file" must point to a publicly accessible `.hApp` file.
    * (Optional) "URL of hosted UI" is the URL which your users will access your UI from.

>     * (Optional) Configuration of memproof/registration code.

Author: No this is not a deploy guide and the content re: how to properly configure registration code / memproof set up is part of a deploy guide. We might revisit this if there is a learning pathway issue but right now I'd rather we prepare a "deploy guide" as a standalone.

To use the WebSDK, import and instantiate it wherever you would normally import and instantiate `@holochain/client`:

```typescript
import WebSdk from '@holo-host/web-sdk';

const client: WebSdk = await WebSdk.connect({
  chaperoneUrl: 'https://chaperone.holo.hosting',
  authFormCustomization: {
    appName: 'example-happ',
  }
});
```

WebSDK documentation can be found [here](https://github.com/Holo-Host/web-sdk#holo-hosting-web-sdk)

#### Sign-up, sign-in, sign-out

Users derive application-specific keys directly in Chaperone using their email and password. This is done directly in the iframe provided by the Chaperone connection manager, which the application developer can customize and must invoke somewhere in their user flow:

```typescript
client.signUp();
client.signIn();
```

> Since AdminWebsocket functionality is handled via Envoy, a service running on Holo hosts that handles the provisioning of hApps, Holo decides whether to show a "sign-up" or "log-in" form based on whether a cell needs to be provisioned for the user.

Reviewer: I don't quite understand this sentence -- is it saying that things normally accessible in the self-hosted context are not accessible in the Holo context because Envoy intercepts some of them? If that's the case, this is (or should be) true of the self-hosted context too -- admin websocket shouldn't be accessible to the UI. So the sentence on line 748 might also describe a call that UI devs shouldn't be making anyway.

(Check my facts; I might be wrong about this)

Author: AdminWebsocket calls aren't exposed to the UI?

This also means that developers cannot assume that calls such as `.agent_info()` are immediately available at first.

Users should also have a means of clearing their keys from local storage using:

```typescript
client.signOut();
```

#### Handling connection state

Since the host conductor is remote, UIs need to handle different connection states appropriately. Chaperone makes connection state information available to the client via events that the UI can listen for and handle. Example:

```typescript
client.on('agent-state', (agent_state: AgentState) => {
  // Handle changes to agent state
});
```

> This generally results in a UI with a much more asynchronous programming pattern, whereas native Holochain UIs tend to be use a more synchronous programming pattern.

Reviewer: So far I've mostly seen native HC UIs programming in async style because it's all about waiting for WebSocket responses. Is there more beyond this that needs to be done -- some sort of mounting/unmounting of all zome calls based on agent state?

Author: So my impression is that in native Holochain you have serial async approach

I.e. 
```
async function foo() -> Promise<Bar> { ... }
const result = await foo

```

In Holo UIs you tend to have more of a listener pattern i.e.

```
async function foo() -> Promise<Void> { ... }

listener.on('event_that_foo_triggers') => { ... }
```

Reviewer: ohh, got it, so you wouldn't use `await` as much in Holo-hosted? Is that because of higher latency / greater risk of disconnection?

### hApp paradigm

hApps are the fundamental unit within Holo. This means:

* A user will have unique agent keys for each hApp, even if multiple hApps share the same DNA.
* For each user, all DNAs/cells within the same hApp will share the same keys.
    * All cells for a hApp are guaranteed to be co-located on the same Holo host.
    * Cloned cells are treated as part of their parent hApp bundle and are thusly also co-located with the other cells in the hApp.
* hApps are deployed onto the Holo network by an agent of the Cloud Console app, who is the designated manager for the hApp.
* hApp managers are responsible for deploying application updates and paying host invoices.

### Multi-tenant conductor

Holo makes use of multi-tenant conductors to provide functionality typically expected of traditional web applications. Each conductor runs applications belonging to multiple users.

> By default, Holo will run read-only cells of deployed applications **under each host's agent** in addition to any user cells. These cells serve two purposes:

Reviewer: ```suggestion
By default, Holo will run **read-only cells** of deployed applications **under each host's agent** in addition to any user cells. These cells serve two purposes:
```

Thought that point deserves extra emphasis.

#### Providing high-uptime nodes

These nodes will send and receive gossip to sustain the network as long as hosts are online. They also ensure that when users connect to their host they already have the latest data available, since Holochain maintains a single local copy of a DHT for all cells of that DNA.

#### Serving data to anonymous users

For some use cases, it's not desirable to require users to create an account, for example, reading an article. Where developers make these use cases available to their users, the data is retrieved by the read-only instance.

> This behavior is available by default. Once a WebSDK client has been instantiated and before a user derives their agent keys, any zome calls will follow this pattern and have read access by default.

Reviewer: How does this work -- how does the stack know which functions are only going to need read access? What cap grant does an anonymous Chaperone connection use? AFAIU this would require an unrestricted grant on all zome functions.

Author: Thanks. This is old. I will update in a separate PR

> If you want to securely disable this functionality, you will need to implement a membrane proof with read-only restrictions. As a less secure alternative, you can:

Reviewer: * Which functionality specifically? the 'free pass to everything' functionality?
* I'm confused by the phrase 'implement a membrane proof with read-only restrictions' -- do you mean a membrane proof validation function that allows the read-only agent's membrane proof paired with a general validation function that denies any source chain writes? Or are source chain writes denied at the conductor level somehow?
* I'm not sure we need to give a less secure alternative -- I'd prefer just giving one (secure) piece of advice to everyone. Thoughts?

Author: Happy to remove the less secure alternative. It's kinda hacky.

Author: I think generally I need to re-write this section

1. Set up your UI lifecycle to call the sign-up/sign-in form and ensure credentials are set up before requesting any data from the DHT, and
2. Ensure that `.signUp({cancellable: false})` or `.signIn({cancellable: false})` are configured.

#### Admin API WebSocket access

> Due to the security implications of multi-tenant conductors, `AdminWebsocket` (and some `AppWebsocket`) functionality is not directly exposed to Holo clients. Instead, this functionality is exposed via Holo WebSDK methods where appropriate.

Reviewer: ```suggestion
Due to the security implications of multi-tenant conductors, `AdminWebsocket` (and some `AppWebsocket`) functionality is not directly exposed to Holo clients. Instead, this functionality is exposed indirectly via Holo WebSDK methods where appropriate.
```

Am I right about that?

But also, as mentioned above, this also should be true of native HC UIs. The fact that it isn't yet is just because the restriction isn't enforced yet.

Author: Should it be true of native HC UIs? Or do you mean "exposed to the end user". Because I mean that someone developing a UI for Holohosting using web-sdk cannot make AdminWebsocket calls.

### Membrane proofs

> Holo strongly recommends membrane proof implementations for production or long-running applications. However, as noted earlier, Holo installs and runs cells of deployed applications under each host’s agent to provide high-uptime DHT nodes and to serve data to anonymous users and Web2 participants who do not have keys.

Reviewer: ```suggestion
Holo strongly recommends membrane proof implementations for production or long-running applications. However, as noted earlier, Holo installs and runs read-only cells of deployed applications under each host’s agent to provide high-uptime DHT nodes and to serve data to anonymous users and Web2 participants who do not have keys. **These cells always have a zero-byte membrane proof.**
```

I've got this right, right? This feels like the important take-home message; if you want to create a restricted membership but still allow the read-only instances to serve public content and maintain DHT resilience, you've got to look for the read-only membrane proof.

(I have separate concerns about the easy forgeability of a zero-byte membrane proof that would allow a non-host access to any DHT they wanted, if the following were implemented in a DNA -- am I wrong about how that works?)

Author: That is correct. I'm wondering if we should clarify that "Holo CAN install..." i.e. you can disable this functionality by simply not providing the zero-byte membrane proof. What do you think?

Re: the easy forgeability, yes that is also something on our radar that we don't have a good answer to currently. Long term the solution is to allow HoloPorts to request membrane proofs from the happ manager's infrastructure directly using some sort of secret.

> Membrane proofs that restrict this read-only access prevent Holo from providing such affordances for your hApp. This may be an acceptable trade-off or desired behaviour. However, If you want to benefit from always-on nodes and read-only access while having a membrane proof, we you will need to configure your membrane proof logic to allow "unpermissioned" read-only access like below:

Reviewer: ```suggestion
DNAs that require membrane proofs restrict this read-only access and thus prevent Holo from providing such affordances for your hApp. This may be an acceptable trade-off or desired behaviour. However, If you want to benefit from always-on nodes and read-only access while having a membrane proof, you will need to configure your membrane proof logic to allow "unpermissioned" read-only access like below:
```

```rust
pub fn is_read_only_proof(mem_proof: &MembraneProof) -> bool {
    let b = mem_proof.bytes();
    b == &[0]
}

> pub fn validate_joining_code(

Reviewer: This function might be misconstrued as an official callback. I wonder if we could demonstrate the use of this function within `validate()`?

It also seems like it needs validation code to deny all source chain writes past the genesis records -- or is this enforced at the conductor level?

Author: @robbiecarlton Can you answer this question?

```rust
pub fn set_cap_tokens() -> ExternResult<()> {
    let mut fn = BTreeSet::new();
    fns.insert((zome_info()?.name, "get_article".into()));
    fns.insert((zome_info()?.name, "get_all_articles".into()));

    let functions = GrantedFunctions::Listed(fns);
    create_cap_grant(CapGrantEntry {
        tag: "".into(),
        access: CapAccess::Unrestricted,
        functions,
    })?;
    Ok(())
}

#[hdk_extern]
fn init(_: ()) -> ExternResult<InitCallbackResult> {
    set_cap_tokens()?;

    Ok(InitCallbackResult::Pass)
}
```

In general, you should not list functions in the above code that _write_ to the source chain, only that read from the source chain.

Finally, if you want to make sure that anonymous agents can't write to their own source chain, you should add the following protection to any function that writes to the source chain:

> ```rust

Reviewer: As mentioned above, I'd recommend moving this restriction to the integrity zome's `validate()` function, so that any records past `InitComplete` are prevented at both the conductor and DHT level and can't be circumvented by simply hot-swapping coordinator zomes. I don't know what this would look like -- how many different types of `Op` you'd have to put this guard on. Probably all of them 🤷🏼‍♂️ The one problem with this is that you'd have to do it for _each_ integrity zome, and some of those zomes might be third-party.

Author: @robbiecarlton This one too

Reviewer 2: Paul's right, this would be more correct.

Reviewer: How to make this performant is a question though -- you don't want to be doing a `must_get_agent_activity` on every single zome call. I suppose it could work like this when validating:

1. If the preceding source chain action is an agent key, require a non-read-only membrane proof preceding it.
2. If the preceding source chain action is anything else and a validator has marked it valid, it's assumed it was also subjected to these two rules, so the action currently being valid must also be writable.

This recursive rule inductively validates a whole chain but only requires a single DHT get per validation. (Of course, it also depends on the honesty of validators.)

Reviewer: Another problem (mentioned in chat, but noting here for visibility) is that this would have to be explicitly coded into the validation function of each integrity zome, which would prevent you from using third-party zomes. All in all, the validation approach is 'more correct' (= secure) if zero-byte memproofs are to be used, but very challenging to make workable in our current composability model -- Integrity zomes are closed to both modification and extension.

    // Read-only membrane proofs are always permitted.
    if is_read_only_membrane_proof(membrane_proof.clone()) {
        return Ok(ValidateCallbackResult::Valid);
    }

    // Your custom membrane proof validation should go here.
}

/// A helper function to check wheter an op should be allowed. In the sample
/// `validate` function, it's used for both entry and link CRUD actions.
fn has_permission_to_write(op: &Op) -> Result<bool, WasmError> {
    let action_type = op.action_type();

    // All non-CRUD actions should be allowed.
    if let ActionType::Dna | ActionType::AgentValidationPkg | ActionType::InitZomesComplete | ActionType::OpenChain | ActionType::CloseChain = action_type {
        return Ok(true);
    }

    // In general, all CRUD actions for a read-only node should be invalid.
    // The only ones that _must_ be valid are actions that create (but don't
    // update) special system entries:
    //
    // * `AgentPubKey`
    // * `CapGrant`
    // * `CapClaim` (optional)
    //
    // When allowing cap grants, we're unable to check whether the grant should
    // be allowed -- that has to happen in the coordinator zome.
    if let ActionType::Create = action_type {

>         if let EntryType::AgentPubKey | EntryType::CapClaim | EntryType::CapGrant = op.entry_data().unwrap().1 {

Reviewer 3: I am always sceptical of using unwrap, especially in the integrity zome.
Would it be better to handle the unwrap?

                return Ok(!is_read_only_membrane_proof(action_data.membrane_proof.clone()));
            }
            (a, _) => {
                // Prior action was neither CRUD nor membrane proof; walk
                // further along the source chain.
                prior_action_hash = a.prev_action().map(|h| h.clone());
            }
        }
    }

    Err(wasm_error!("Undefined behavior; this place should be unreachable"))
}

#[hdk_extern]
pub fn genesis_self_check(
    data: GenesisSelfCheckData,
) -> ExternResult<ValidateCallbackResult> {
    validate_membrane_proof(data.membrane_proof.clone())
}

#[hdk_extern]
pub fn validate(op: Op) -> ExternResult<ValidateCallbackResult> {
    // Here the membrane proof itself is being validated.
    // The AgentValidationPkg action produces two different ops that
    // contain a membrane proof; let's check both of them.
    if let Action::AgentValidationPkg(action) = match &op {
        Op::StoreRecord(store_record) => Some(store_record.record.signed_action.hashed.content.clone()),
        Op::RegisterAgentActivity(register_agent_activity) => Some(register_agent_activity.action.hashed.content.clone()),
        _ => None

>     }.unwrap() {

Reviewer: crap, this will regularly panic now, I bet. Just about to push a fix.

#### Providing high-uptime nodes

These nodes will send and receive gossip to sustain the network as long as hosts are online. They also ensure that when users connect to their host they already have the latest data available, since Holochain maintains a single local copy of a DHT for all cells of that DNA.

#### Serving data to anonymous users

For some use cases, it's not desirable to require users to create an account, for example, reading an article. Where developers make these use cases available to their users, the data is retrieved by a read-only cell.

This behavior isn't available by default; you'll have to implement functionality to both give anonymous users access to read-only functions and prevent them from making writes. This is described in the [next section](#implementing-read-only-cells).

### A note on admin API WebSocket access

Due to the security implications of multi-tenant conductors, `AdminWebsocket` (and some `AppWebsocket`) functionality is not directly exposed to Holo clients. Instead, this functionality is exposed indirectly via Holo WebSDK methods where appropriate.

### Implementing read-only cells

> In order to provide high-uptime nodes and anonymous access via read-only cells while keeping your hApp safe from defacement, you'll need to restrict what a read-only cell can do. A read-only cell is **read-only merely by convention**; that is, because you're checking for read-only cells and handling writes differently.
> 
> The most secure way to restrict access to a hApp's data is with [**membrane proofs**](/references/glossary/#membrane-proof). A membrane proof is supplied at installation time and acts like an ID card, allowing the hApp to admit valid members only. However, this also prevents Holo hosting devices from provisioning read-only cells.
> 
> This is because read-only cells always use a zero-byte membrane proof. This is essentially the same as no membrane proof, so this is only appropriate for hApps that contain data that should be publicly readable but not anonymously writable. (Note that, in the future, the production Holo hosting network may permit you to implement logic that provisions membrane proofs to read-only nodes, allowing you to use Holo for high-uptime while preventing anonymous read access, but this work is currently out of scope.)
> 
> To implement logic that allows read-only cells to join a DNA's network but not write data, your integrity zome's validation function should check the author's membrane proof, then allow them to join the network but reject CRUD operations if the proof is zero bytes.

Author: ```suggestion
In order to provide high-uptime nodes and anonymous hosted user access via read-only cells while keeping your hApp safe from defacement, you'll need to restrict what a read-only cell can do. A read-only cell is **read-only merely by convention**; that is, because you're checking for read-only cells and handling writes differently.

The most secure way to restrict access to a hApp's data is with [**membrane proofs**](/references/glossary/#membrane-proof). A membrane proof is supplied at installation time and acts like an ID card, allowing the hApp to admit valid members only. However, this prevents Holo hosting devices from provisioning read-only cells unless explicit affordances are made.

The Holo Network expects and uses a zero-byte membrane proof for read-only cells. A publicly known membrane proof without restriction is essentially the same as no membrane proof. To implement logic to allow read-only cells to join a DNA's network but not write data, your integrity zome's validation function should check the author's membrane proof, then allow them to join the network but reject CRUD operations if the proof is zero bytes.

(Note that, in the future, the production Holo hosting network may permit you to implement logic that provisions membrane proofs to read-only nodes, allowing you to use Holo for high-uptime while preventing anonymous read access.)
```

Reviewer: brilliant, thanks!

> The following code shows an example of how to do both of these things. This isn't necessarily the most performant solution, as it requires a validator to retrieve the author's entire source chain (excluding entry data). A better-performing solution might rely on [inductive validation](/references/glossary/#inductive-validation) of the author's source chain.

Author: Why can't we do the better-performing one? If people are basically going to copy and paste this stuff we should write something we are happy with. Especially since changing this will be an integrity zome change.

Reviewer: done, except for review on the code of the better-performing one.

    // The only ones that _must_ be valid are actions that operate on special
    // system entries:
    //
    // * `AgentPubKey`
    // * `CapGrant`
    // * `CapClaim` (optional)
    //
    // Your app may also want to allow certain public or private CRUD actions,
    // such as anonymous pageview counters for blog posts. Keep in mind the
    // risk of anonymous spamming though!

    // Find the membrane proof for the op's author.
    // Start by getting the author's source chain.
    let chain_top = op.prev_action().unwrap();
    let agent_id = op.author();
    let chain_filter = ChainFilter::new(chain_top.clone());
    let chain = must_get_agent_activity(agent_id.clone(), chain_filter)?;

    // Now find the membrane proof at the beginning of the chain.
    let membrane_proof = chain
        .iter()
        .find(|record| match record.action.hashed.content {
            Action::AgentValidationPkg(_) => true,
            _ => false
        })
        .map(|record| match &record.action.hashed.content {
            Action::AgentValidationPkg(action) => action.membrane_proof.clone(),
            _ => None
        });

>     Ok(!is_read_only_membrane_proof(membrane_proof.unwrap()))

Reviewer 3: The unwrap in this statement can panic because the agent could have no mem-proof, right?

}

#[hdk_extern]
pub fn genesis_self_check(
    data: GenesisSelfCheckData,
) -> ExternResult<ValidateCallbackResult> {
    validate_membrane_proof(data.membrane_proof.clone())
}

#[hdk_extern]
pub fn validate(op: Op) -> ExternResult<ValidateCallbackResult> {
    match op.action_type() {
        hdi::prelude::ActionType::AgentValidationPkg => {
            // Here the membrane proof itself is being validated.
            // The AgentValidationPkg action produces two different ops that
            // contain a membrane proof; let's check both of them.
            if let Action::AgentValidationPkg(action) = match &op {
                Op::StoreRecord(store_record) => Some(store_record.record.signed_action.hashed.content.clone()),
                Op::RegisterAgentActivity(register_agent_activity) => Some(register_agent_activity.action.hashed.content.clone()),
                _ => None
            }.unwrap() {
                return  validate_membrane_proof(action.membrane_proof.clone());

>             }

Reviewer 3: This if statement needs an else

Reviewer: This is just to get around the awkward fact that I've already checked the action type so I know the op's associated action is `AgentValidationPkg`. All I'm doing here is `if let` matching on a condition I know must be true. I've simplified it in a later commit.

        },

        hdi::prelude::ActionType::Create | hdi::prelude::ActionType::Update | hdi::prelude::ActionType::Delete => {
            // If not, deny writes to read-only cells.

>             match op.entry_data().unwrap() {

Reviewer 3: The unwrap can panic too right? because the initial creates do not have entries just actions

Reviewer: Hm, yeah, looks like `Action::entry_data()` is guaranteed to return something meaningful for new-entry actions (even for private entries, and even for ops that don't carry entry bytes), but not for delete-entry actions. Nice catch, thank you.

}
```

#### Anonymous zome function access

Every function call to a coordinator zome must be signed, and a call will only be successful if there's a [capability grant](/references/glossary/#capability-grant) for it. Capability grants can be restricted to a particular keypair, a particular capability token, or unrestricted.

Normally, when a Holo agent is logged in, or the user is running the hApp on their own machine, the keypair used to sign function calls is the same as the keypair used to author data in the cell. This is called the [author grant](/concepts/8_calls_capabilities/#author-grant), and it's automatically privileged to call every function. For an anonymous agent accessing a read-only instance, however, this is not true. For them to be able to make a function call, you need to also create an unrestricted capability grant for that function.

```rust
pub fn set_cap_tokens() -> ExternResult<()> {
    let mut fn = BTreeSet::new();
    fns.insert((zome_info()?.name, "get_article".into()));
    fns.insert((zome_info()?.name, "get_all_articles".into()));

    let functions = GrantedFunctions::Listed(fns);
    create_cap_grant(CapGrantEntry {

>         tag: "".into(),

Reviewer 3: ```suggestion
        tag: "read-only".into(),
```

Reviewer: nice idea.

        access: CapAccess::Unrestricted,
        functions,

>     })?;

Reviewer 3: This create is going to require write access, and with the above validation rule, it will block the read-only mem-proof from committing this grant right?

      IS_HOLO: ['true', '1', 't'].includes(import.meta.env.VITE_APP_IS_HOLO?.toLowerCase()), // This will be passed later as an env variable
    };
  },
```

> The WebSDK client loads and connects to an iframe called Chaperone. Chaperone is a key and connection manager to the Holo network and is invisible outside a few specific cases. Let's initialise this client in the Vue `mounted()` lifecycle hook.

Reviewer: ```suggestion
The WebSDK client loads and connects to an iframe called Chaperone. Chaperone is a key and connection manager to the Holo network and is invisible outside a few specific cases. Let's initialize this client in the Vue `mounted()` lifecycle hook.
```

I am so sorry; we've standardised on American spelling. It pains me too. (my spell checker tells me 'standardised' is spelled wrong and I'm stubbornly ignoring it. But I would write 'standardized' in the dev docs)
